cpp/tests/stream_compaction/distinct_count_tests.cpp

```diff
@@ -0,0 +1,310 @@
+
+  auto const new_end      = std::unique(input.begin(), input.end());
+  auto const gold_ordered = new_end - input.begin() - 5;
+  EXPECT_EQ(gold_ordered,
+            cudf::distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_VALID));
+}
+
+TEST_F(DistinctCount, WithNansAndNull)
+{
+  using T = float;
+
+  std::vector<T> input               = {1,  3,  NAN, 70, 31,  1, 8,   2, 0, 70, 1,
+                          70, 10, 40,  31, NAN, 0, NAN, 8, 5, 70};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1,
+                                        0, 1, 1, 1, 1, 1, 1, 1, 1, 0};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{input.begin(), input.end(), valid.begin()};
+
+  cudf::size_type const expected = std::set<T>(input.begin(), input.end()).size();
+  EXPECT_EQ(
+    expected + 1,  // +1 since `NAN` is not in std::set
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_VALID));
+
+  auto const new_end      = std::unique(input.begin(), input.end());
+  auto const gold_ordered = new_end - input.begin();
+  EXPECT_EQ(gold_ordered,
+            cudf::distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_VALID));
+}
+
+TEST_F(DistinctCount, WithNansOnly)
+{
+  using T = float;
+
+  std::vector<T> input               = {1, 3, NAN, 70, 31};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 1, 1};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{input.begin(), input.end(), valid.begin()};
+
+  constexpr auto expected = 5;
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_VALID));
+  EXPECT_EQ(expected,
+            cudf::distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_VALID));
+}
+
+TEST_F(DistinctCount, NansAsNullWithNoNull)
+{
+  using T = float;
+
+  std::vector<T> input               = {1, 3, NAN, 70, 31};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 1, 1};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{input.begin(), input.end(), valid.begin()};
+
+  constexpr auto expected = 5;
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_NULL));
+  EXPECT_EQ(expected,
+            cudf::distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_NULL));
+}
+
+TEST_F(DistinctCount, NansAsNullWithNull)
+{
+  using T = float;
+
+  std::vector<T> input               = {1, 3, NAN, 70, 31};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 0, 1};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{input.begin(), input.end(), valid.begin()};
+
+  constexpr auto expected = 4;
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_NULL));
+  EXPECT_EQ(expected,
+            cudf::distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_NULL));
+}
+
+TEST_F(DistinctCount, NansAsNullWithIgnoreNull)
```

Comment:
The test coverage could be stronger if we added explicit tests like `NansAsNullWithIgnoreNull` but the input data is _only_ NaNs, like `{NAN, NAN, NAN}`. Similar for nulls and with other nan/null handling parameters.

Reply:
> The test coverage could be stronger

Can I ask why is it stronger to test a pure `NaN` array?

Reply:
> Can I ask why is it stronger to test a pure `NaN` array?

There's some special handling of `NaN`s and nulls in this section of code, where the count is manually adjusted: https://github.com/rapidsai/cudf/blob/57ff6f55b9fd44e8a8e10282d3f95d5f38e299ef/cpp/src/stream_compaction/distinct_count.cu#L149-L167

I'd like the tests to cover edge cases like "all nans" or "all nulls" that might be mishandled during a refactor of this logic.

Reply:
Done.